### PR TITLE
[Sprint 9 / Enhancement] Refactor for Akushon

### PR DESCRIPTION
**Pull request by maroqijalil**
## Description

Add complete algorithm of `get_joints_server` for serving current joints value, especially for Akushon package.

## Type of Change

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause the existing functionality to not work as expected)

## How Has This Been Tested?

- [ ] New unit tests added.
- [ ] Manual tested.

## Checklist:

- [x] Using Branch Name Convention
    - `feature/JIRA-ID-SHORT-DESCRIPTION` if has a JIRA ticket
    - `enhancement/SHORT-DESCRIPTION` if has/has no JIRA ticket and contain enhancement
    - `hotfix/SHORT-DESCRIPTION` if the change doesn't need to be tested (urgent)
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made the documentation for the corresponding changes.

### CODE REVIEW

File: src/tachimawari/joint/node/joint_manager.cpp [hunk 1]
```diff
@@ -24,15 +24,56 @@
...
+    }
+  }
+
+  is_joints_uptodate = true;
+}
+
+std::vector<Joint> JointManager::get_current_joints()
+{
+  if (!is_joints_uptodate) {
+    if (control_manager->bulk_read_packet(current_joints)) {
```
Comment (segara2410) [inline src/tachimawari/joint/node/joint_manager.cpp]: change these nested ifs to return earlier or use && instead

File: include/tachimawari/joint/node/joint_manager.hpp [hunk 2]
```diff
@@ -36,13 +36,19 @@ class JointManager
...
 
+  std::vector<Joint> get_current_joints();
+
 private:
+  void update_current_joints(const std::vector<Joint> & joints);
+
   std::shared_ptr<tachimawari::control::ControlManager> control_manager;
+
+  std::vector<Joint> current_joints;
+  bool is_joints_uptodate;
```
Comment (segara2410) [inline include/tachimawari/joint/node/joint_manager.hpp]: kinda grammatically incorrect, you can use `is_each_joint_updated`

File: src/tachimawari/joint/node/joint_manager.cpp [hunk 3]
```diff
@@ -24,15 +24,56 @@
...
+          position = (current_position == -1) ? 0.0 : current_position;
+        }
+
+        current_joints[index].set_position(position);
+      }
+    }
+  }
+
+  return current_joints;
 }
```
Comment (threeal) [inline src/tachimawari/joint/node/joint_manager.cpp]: seems it's fine to return as `const std::vector<Joint> &` isntead.

File: src/tachimawari/joint/node/joint_manager.cpp [hunk 4]
```diff
@@ -24,15 +24,56 @@
...
+        float position = 0.0;
+
+        if (control_manager->get_protocol_version() == 1.0) {
+          int current_position = control_manager->get_bulk_data(
+            current_joints[index].get_id(), protocol_1::MX28Address::PRESENT_POSITION_L, 2);
+          position = (current_position == -1) ? 0.0 : current_position;
+        }
+
+        current_joints[index].set_position(position);
+      }
```
Comment (threeal) [inline src/tachimawari/joint/node/joint_manager.cpp]: As `index` is only used to access the list, it seems you could do this instead:
```c++
for (auto & current_joint : current_joints) {
  ...
  current_joint.set_position(position);
}
```

see this too:
https://github.com/ichiro-its/tachimawari/blob/d13949f04f67fe2ca6135c4ae7e75406c5c42fce/src/tachimawari/joint/node/joint_manager.cpp#L46-L52

File: src/tachimawari/joint/node/joint_node.cpp [hunk 5]
```diff
@@ -56,6 +56,20 @@ JointNode::JointNode(rclcpp::Node::SharedPtr node, std::shared_ptr<JointManager>
...
+          std::vector<Joint> joints_msg;
+          for (const auto & joint : this->joint_manager->get_current_joints()) {
+            auto joint_msg = Joint();
+            joint_msg.id = joint.get_id();
+            joint_msg.position = joint.get_position();
+
+            joints_msg.push_back(joint_msg);
+          }
+
+          response->joints = joints_msg;
```
Comment (threeal) [inline src/tachimawari/joint/node/joint_node.cpp]: if performance is important, you could do this instead:
```c++
auto current_joints = this->joint_manager->get_current_joints(); // or const auto &, depend on the return type
auto & joints = response->joints;
joints.resize(current_joints.size());
for (size_t i = 0; i < joints.size() && i < current_joints.size(); ++i) {
  joints[i].id = current_joints[i].get_id();
  ...
}
```

**Review by threeal — APPROVED** [review]
looks good though 👍

File: src/tachimawari/joint/node/joint_manager.cpp (hunk 1 above)
Comment (maroqijalil) [reply src/tachimawari/joint/node/joint_manager.cpp]: done

File: include/tachimawari/joint/node/joint_manager.hpp (hunk 2 above)
Comment (maroqijalil) [reply include/tachimawari/joint/node/joint_manager.hpp]: done

File: src/tachimawari/joint/node/joint_manager.cpp (hunk 3 above)
Comment (maroqijalil) [reply src/tachimawari/joint/node/joint_manager.cpp]: done

File: src/tachimawari/joint/node/joint_manager.cpp (hunk 4 above)
Comment (maroqijalil) [reply src/tachimawari/joint/node/joint_manager.cpp]: done

File: src/tachimawari/joint/node/joint_node.cpp (hunk 5 above)
Comment (maroqijalil) [reply src/tachimawari/joint/node/joint_node.cpp]: done

**Review by segara2410 — APPROVED** [review]

**Pull request merged**